mongo/integration/unified/matches_test.go

```diff
@@ -323,3 +336,175 @@ func newMatchingError(keyPath, msg string, args ...interface{}) error {
+		assert.Nil(t, err, "UnmarshalExtJSON error: %v", err)
+		return val
+	}
+	marshalValue := func(t *testing.T, val interface{}) bson.RawValue {
+		t.Helper()
+
+		valType, data, err := bson.MarshalValue(val)
+		assert.Nil(t, err, "MarshalValue error: %v", err)
+		return bson.RawValue{
+			Type:  valType,
+			Value: data,
+		}
+	}
+
+	assertMatches := func(t *testing.T, expected, actual bson.RawValue, shouldMatch bool) {
+		t.Helper()
+
+		err := VerifyValuesMatch(ctx, expected, actual, true)
+		if shouldMatch {
+			assert.Nil(t, err, "expected values to match, but got comparison error %v", err)
+			return
+		}
+		assert.NotNil(t, err, "expected values to not match, but got no error")
+	}
+
+	t.Run("documents with extra keys allowed", func(t *testing.T) {
+		expectedDoc := `{"x": 1, "y": {"a": 1, "b": 2}}`
+		expectedVal := unmarshalExtJSONValue(t, expectedDoc)
+
+		extraKeysAtRoot := `{"x": 1, "y": {"a": 1, "b": 2}, "z": 3}`
+		extraKeysInEmbedded := `{"x": 1, "y": {"a": 1, "b": 2, "c": 3}}`
+
+		testCases := []struct {
+			name    string
+			actual  string
+			matches bool
+		}{
+			{"exact match", expectedDoc, true},
+			{"extra keys allowed at root-level", extraKeysAtRoot, true},
+			{"incorrect type for y", `{"x": 1, "y": 2}`, false},
+			{"extra keys prohibited in embedded documents", extraKeysInEmbedded, false},
+		}
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				assertMatches(t, expectedVal, unmarshalExtJSONValue(t, tc.actual), tc.matches)
+			})
+		}
+	})
+	t.Run("documents with extra keys not allowed", func(t *testing.T) {
+		expected := unmarshalExtJSONValue(t, `{"x": 1}`)
+		actual := unmarshalExtJSONValue(t, `{"x": 1, "y": 1}`)
+		err := VerifyValuesMatch(ctx, expected, actual, false)
+		assert.NotNil(t, err, "expected values to not match, but got no error")
+	})
+	t.Run("exists operator", func(t *testing.T) {
+		rootExists := unmarshalExtJSONValue(t, `{"x": {"$$exists": true}}`)
+		rootNotExists := unmarshalExtJSONValue(t, `{"x": {"$$exists": false}}`)
+		embeddedExists := unmarshalExtJSONValue(t, `{"x": {"y": {"$$exists": true}}}`)
+		embeddedNotExists := unmarshalExtJSONValue(t, `{"x": {"y": {"$$exists": false}}}`)
+
+		testCases := []struct {
+			name     string
+			expected bson.RawValue
+			actual   string
+			matches  bool
+		}{
+			{"root - should exist and does", rootExists, `{"x": 1}`, true},
+			{"root - should exist and does not", rootExists, `{"y": 1}`, false},
+			{"root - should not exist and does", rootNotExists, `{"x": 1}`, false},
+			{"root - should not exist and does not", rootNotExists, `{"y": 1}`, true},
+			{"embedded - should exist and does", embeddedExists, `{"x": {"y": 1}}`, true},
+			{"embedded - should exist and does not", embeddedExists, `{"x": {}}`, false},
+			{"embedded - should not exist and does", embeddedNotExists, `{"x": {"y": 1}}`, false},
+			{"embedded - should not exist and does not", embeddedNotExists, `{"x": {}}`, true}, // TODO
```

Comment:
Why TODO?

Reply:
It was previously marked TODO because it was failing due to a bug. Removed the TODO.